### ISSUE TITLE
Obtaining data in timeseries format fails with "this.datapoints is undefined (TypeError)."

### ISSUE THREAD

**Issue by perosb** [hiveeyes/grafana-worldmap-panel#3]
Hi,

swapped from 0.2 to try out the 0.3 dev version.
I am using Prometheus with a custom json endpoint for locations.
![image](https://user-images.githubusercontent.com/982752/58633711-8c0a8000-82e9-11e9-9833-d91427ab62af.png)

![image](https://user-images.githubusercontent.com/982752/58633681-76955600-82e9-11e9-97ff-bd255cb2fd11.png)

**Comment by perosb** [comment]
Not sure if I'm reading this right but this looks strange:

![image](https://user-images.githubusercontent.com/982752/58635361-f9201480-82ed-11e9-8a9c-8adb4f88a3e6.png)
seriesData is PluginSettings and settings contains datapoints
![image](https://user-images.githubusercontent.com/982752/58635384-0e953e80-82ee-11e9-8bc0-44226e9963d1.png)

**Comment by amotl** [comment]
Dear Per,

thanks for finding us and giving this a try. We are happy to improve the last mile of the refactoring in order to make this production ready as mentioned with [1] and especially improve error handling and reporting. I would like to quickly go more into the details of your observations - without actually using Prometheus yet - so please bear with me.

So,
1. We recently found our refactoring somehow broke feeding data in timeseries format into the Worldmap Panel machinery. We discovered this when rolling out the 0.3.0-dev Panel to one of our servers where we feed data from InfluxDB in timeseries format instead of table format. We will have to fix this!

2. After some short research, we found two references [2,3] where @daniellee pointed out to use Prometheus with **Table Format**. Saying that, we also already found that not all data sources will work with all formats, so you have to choose the right format for your data. After discovering this, we tried hard to improve error handling and reporting [4] in order to at least give the user a better chance to find out what is actually going on.

So, we a) would like to ask you to try your Prometheus data source with table format and b) will be ready to improve the handling of Prometheus data in timeseries format, especially when you will tell us that it worked before™ ;].

Thanks again for reporting this, testing the overhauled Worldmap with the most popular TSDBs out there is important to us, but as we are usually only running InfluxDB, we need the support of the community.

With kind regards,
Andreas.

[1]
> Please be aware that we primarily use InfluxDB, so we would be especially happy to hear if this still works for people using different data sources.
-- https://community.grafana.com/t/giving-the-grafana-worldmap-panel-some-love/17210
[2] https://community.grafana.com/t/prometheus-grafana-worldmap-not-working/1997
[3] https://community.grafana.com/t/worldmap-panel-with-prometheus-metrics/5137
[4] https://community.hiveeyes.org/t/grafana-worldmap-panel-0-3-0-dev-series/1824/23

**Comment by perosb** [comment]
Thanks for the long and detailed response.
I changed to table format and also changed the mapping to table+json.
Now I'm getting:
![image](https://user-images.githubusercontent.com/982752/58660733-68195f80-8326-11e9-8999-ff2598f67fd8.png)


Not sure how this should work but it tries to get lat/lng from the prometheus query? That query should only contain the metric+value right?
And the lat/lng should come from the JSON api call.
In my case then it sets all "key, row #321, to undefined_undefined.
It looks up the correct location on #329 but never uses the lat/lng, only the labels/name?

![image](https://user-images.githubusercontent.com/982752/58660181-20de9f00-8325-11e9-9f18-071a4cb64f8f.png)

**Comment by perosb** [comment]
I suspect I'm missing a "Table, resolve locaton fro custom JSON" option?

![image](https://user-images.githubusercontent.com/982752/58661761-f4c51d00-8328-11e9-8861-ea62fd6af7c3.png)

**Comment by amotl** [comment]
Dear Per,

## Investigation
We are just now starting to ramp up our **worldmap-with-timeseries-format** skills (we only used InfluxDB/Table-Format while working on 0.3.0-dev) and we learned - as you also might have discovered earlier than us - that
> Not sure how this should work but it tries to get lat/lng from the prometheus query? That query should only contain the metric+value right?

So, 
> And the lat/lng should come from the JSON api call.

yes!

## Let's summarize
- When using the table format, location information for the points on the map is coming from the primary data source itself and **may** become enriched by out-of-band JSON data.
- When using timeseries format, location information **must** be obtained from out-of-band JSON data, as the primary data source itself does not carry any information about that at all.

We haven't been exactly aware of this very difference until just now, so to make sure we are on the same page: Do you agree with that?

## Conclusion
> I suspect I'm missing a "Table, resolve location from custom JSON" option?

When using Table format, that would require such an operation mode, indeed. However, as I am currently getting into the details, using the Timeseries format obviously would be the better option then, while obtaining location information from JSON by key-lookup as originally intended.

While being at it: Are you able to share a little excerpt from the very JSON file you have been using when trying the timeseries-format option at first hand just to make sure this has the appropriate structure and values?

## Bottom line
So, we conclude the primary thing we learned is that

![image](https://user-images.githubusercontent.com/453543/58665308-db27d380-8330-11e9-897a-83ea12975728.png)

is the actual bug we already recognized and which should be resolved. We are currently working on it and humbly ask for a bit of patience to get things right including eventually adjusting the error reporting for the timeseries-format operation mode which we haven't touched yet while working on the refactoring.

Thanks again for bringing this up. We are looking forward to make Prometheus-based timeseries-format data work, if things turn out well and we are on the right path.

With kind regards,
Andreas.

**Comment by amotl** [comment]
Just for the records. What you have been observing is probably the very same thing we recognized the other day.

> We recently found on [1] our refactoring somehow broke feeding data in timeseries format into the Worldmap Panel machinery. We discovered this when rolling out the 0.3.0-dev Panel to one of our servers [to a staging-like Grafana 6 instance] where we feed data from InfluxDB in timeseries format instead of table format. We will have to fix this!

![image](https://user-images.githubusercontent.com/453543/58666065-a157cc80-8332-11e9-87cb-e0c590d9f7f2.png)
Left: Timeseries format. Right: Table format. Both InfluxDB.

[1] https://weather.hiveeyes.org/grafana6/d/NP0wTOtmk/weather-hiveeyes-org

**Comment by perosb** [comment]
The  prometheus timeseries returns:

```{
   "status":"success",
   "data":{
      "resultType":"matrix",
      "result":[
         {
            "metric":{
               "colo":"AKL"
            },
            "values":[
               [
                  1559217600,
                  "1"
               ],
               [
                  1559260800,
                  "2.4"
               ]
            ]
         },
         {
            "metric":{
               "colo":"AMS"
            },
            "values":[
               [
                  1559217600,
                  "2.1052631578947367"
               ],
               [
                  1559260800,
                  "0"
               ]
            ]
         }
      ]
   }
}
```

The location data is the "standard" format:


```[
   {
      "key":"ADJ",
      "latitude":31.972700119018555,
      "longitude":35.991600036621094,
      "name":"Amman",
      "country":"Jordan"
   },
   {
      "key":"AQJ",
      "latitude":29.611600875854492,
      "longitude":35.018100738525391,
      "name":"Aqaba",
      "country":"Jordan"
   },
   {
      "key":"AUH",
      "latitude":24.433000564575195,
      "longitude":54.651100158691406,
      "name":"Abu Dhabi",
      "country":"United Arab Emirates"
   }
]
```

Thank you!

**Comment by amotl** [comment]
Dear Per,

we have been able to make running data in timeseries format through the Worldmap Panel work again. At least, we can confirm it works again on [1] with InfluxDB.

The current release is `0.3.0-dev8` and can be installed by invoking
```
grafana-cli --pluginUrl https://github.com/hiveeyes/grafana-worldmap-panel/archive/0.3.0-dev8.zip plugins install grafana-worldmap-panel
```

Happy mapping and with kind regards,
Andreas.

[1] https://weather.hiveeyes.org/grafana6/d/NP0wTOtmk/weather-hiveeyes-org

**Comment by perosb** [comment]
Very good news @amotl 
Thank you so much. It works like a charm!
